examples/se_2_3_localization.cpp

```diff
@@ -0,0 +1,405 @@
+ * - a is the frame acceleration in mixed-representation (equivalent to p_doubledot in continuous time).
+ * - omega_b as the angular velocity of the robot expressed in the robot frame
+ *
+ * The kinematic equations (1) can be written as,
+ * p <-- p + v dt + 0.5 a dt^2
+ * R <-- R Exp_SO3(omega_b)
+ * v <-- v + a dt
+ *
+ * We would like to express these equations in the form,
+ * X <-- X * Exp(u)
+ * where, X \in SE_2(3), u \in R^9 and u_hat \in se_2(3)
+ * Note that here input vector u is expressed in the local frame (robot frame).
+ *
+ * The exponential mapping of SE_2(3) is defined as,
+ * for u = [u_p, u_w, u_v]
+ * Exp(u) = | Exp_SO3(u_w)   JlSO3(u_w) u_p   JlSO3(u_w) u_v |
+ *          | 0    0    0                 1                0 |
+ *          | 0    0    0                 0                1 |
+ * where, JlSO3 is the left Jacobian of the SO(3) group.
+ *
+ * Please see the Appendix C of the paper "A micro Lie theory for state estimation in robotics",
+ * for the definition of the left Jacobian of SO(3).
+ * Please see the Appendix D of the paper, for the definition of Exp map for SE(3).
+ * The Exp map of SE_2(3) is a simple extension from the Exp map of SE(3).
+ * Also, please refer to Example 7 of the paper to understand when and how the left Jacobian of SO(3)
+ * appears in the definitions of Exp maps. The Example 7 illustrates the scenario for SE(3).
+ * We use a direct extension here for SE_2(3).
+ * One can arrive to such a definition by following the convergent Taylor's series expansion
+ * for the matrix exponential of the Lie algebra element (Equation 16 of the paper).
+ *
+ * As a result of X <-- X * Exp(u), we get (2)
+ * p <-- p + R JlSO3(u_w) u_p
+ * R <-- R Exp_SO3(u_w)
+ * v <-- v + R JlSO3(u_w) u_v
+ *
+ * We would like to draw a relationship between the sets of equations (1) and (2).
+ * Additionally, we would also like to make use of the IMU measurements as exogeneous inputs.
+ * Considering R.T as the transpose of R, the IMU measurements are modeled as,
+ *    - linear accelerometer measurements alpha = R.T (a - g) + w_acc
+ *    - gyroscope measurements omega = omega_b + w_omega
+ * Note that the IMU measurements are expressed in the IMU frame (coincides with the robot frame - assumption).
+ * The IMU measurements are corrupted by noise,
+ *    - w_omega is the additive white noise affecting the gyroscope measurements
+ *    - w_acc is the additive white noise affecting the linear accelerometer measurements
+ * It must be noted that we do not consider IMU biases in the IMU measurement model in this example.
+ *
+ * Taking into account all of the above considerations, the exogenous input vector u (3) becomes,
+ *   u = (u_p, u_w, u_v) where,
+ *   u_w = omega dt
+ *   Jinv = inv(JlSO3(u_w)) is the left Jacobian inverse of SO(3) Lie group
+ *   u_p = Jinv * (R.T v dt + 0.5 dt^2 (alpha + R.T g))
+ *   u_v = Jinv * ((alpha + R.T g) dt)
+ *
```

Comment:
I think here there is some unnecessary confusion.

Let us consider `u = [u_w, u_p, u_v]` angle, position, velocity, in se_2(3).

If one defines the exponential map,

 `Exp(u) = [R(u_w), Jl(u_w)*u_p, Jl(u_w)*u_v ; 0 1 0 ; 0 0 1]`

then the motion integration is simply

 ` X <-- X * Exp(u)`

Now the question is how to define `u`.

If you use Jl.inv in the definition of `u`, then you cancel the effect that these Jacobians have on propagation. This means that you get to the same place as if we have defined the Exp as

 `Exp(u) = [R(u_w), u_p, u_v ; 0 1 0 ; 0 0 1]`            (4)

and having those `Jl` and then those other `Jl.inv` is just pointless.

The issue is that you want to have `u` defined without the Jl.inv:

```
u_w = omega*dt = th
u_p = R.T * v * dt + .5 * (alpha + R.T * g) * dt^2        (5)
u_v = (alpha * R.T * g) * dt
```
leading to the following evolution (5):

```
 p <-- p + Jl * v * dt + 0.5 * Jl * a * dt^2
 R <-- R Exp_SO3(omega_b)                                (6)
 v <-- v + Jl * a * dt
```

which is not equal to (1).

The thing is that this new expression is better than (1) !!! So forget about (1) and do instead:

* compute u as in (5)
* integrate `X <-- X * Exp(u)` as in (2), with Exp() defined as in (4)

Please read Section III.C of the paper we wrote on IMU Lie group for a comment on the differences between (1) and (6) -- it is slightly different here than in the paper, but the concepts are valid.

As I said earlier, both methods converge when dt --> 0. Since we use very small dt, there is not going to be a noticeable difference. But, the approach proposed here is more elegant since it is more accurate and also simpler.

Reply:
Notice that things can be done differently. But the key here is that we want to illustrate Lie theory at its best. And this approach here is the cleanest possible.

The approach in (1) considers R3xSO3xR3 instead of SE23, in the sense explored in Example 7 in the Lie paper. On the contrary, (6) is properly a SE23 approach.

Reply:
Maybe I made some mistake, especially regarding the `R.T` details in some formulas. If you see things do not fit, please do not hesitate to come back to me for further analysis.

Reply:
> Please read Section III.C of the paper we wrote on IMU Lie group for a comment on the differences between (1) and (6) -- it is slightly different here than in the paper, but the concepts are valid.

> Notice that things can be done differently. But the key here is that we want to illustrate Lie theory at its best. And this approach here is the cleanest possible.

> The approach in (1) considers R3xSO3xR3 instead of SE23, in the sense explored in Example 7 in the Lie paper. On the contrary, (6) is properly a SE23 approach.

I see. Thanks for noticing this. 

I will go through the formula again and give this a thought to gain some clarity on this and make the necessary changes. 

I shall go through the paper in detail as well.